### ISSUE TITLE
Client certificate rotation?

### ISSUE THREAD

**Issue by AntonNeld** [brianc/node-postgres#2618]
We are using client certificate authentication toward Postgres, and this certificate is rotated pretty often. I would like to switch it out for a new certificate without drastic measures like recreating the Pool. We are using the Javascript version of the library, not the native bindings.

Is there a way to do this?

(Since the 'ssl' object is saved by reference instead of copied, it seems like this could be possible by keeping an external reference to it and modifying it in-place. But this has the feel of an implementation detail, so I'm not sure if I can count on this way not breaking without warning.)

**Comment by brianc** [comment]
Hey @AntonNeld - good question!  So, checking the docs on the `[tls](https://nodejs.org/api/tls.html#tls_certificate_object)` module in node I don't see a way to rotate a certificate there.  Is there a method on the ssl socket to change/rotate a certificate?

**Comment by sehrope** [comment]
I don't think you can rotate it on an established connection. It'd be handled once in the beginning as it's part of the TLS handshake.

IIUC, this is about being able to rotate the certificate in subsequent connections without bringing down the node app. If the TLS settings are modified or they're pointing at a local file that is modified out of band, then new connections established would use those new settings. Combined with a max lifetime for the pool or a "SIGHUP" handler to flush the pool, you could rotate things without any downtime.

Each connection gets it's own copy of any ssl settings when the connection is created which is cloned off the pool's copy: https://github.com/brianc/node-postgres/blob/master/packages/pg/lib/connection.js#L83

So I think that modifying the pools copy at runtime should work for now, though I wouldn't recommend that as a long term approach.

To support something like this I think we should expand the Pool's constructor API to:

1. Split out pool options (ex: idle time or max size) from connection properties (host, port, etc, ...)
2. Allow the user to specify a function that returns the connection properties rather than a fixed object. The pool would invoke that function to generate the connection properties at runtime each time it creates a physical connection. This would be similar to how the dynamic passwords work but for the entire connection config.
